mesonbuild/mcompile.py

```diff
@@ -0,0 +1,127 @@
+# Copyright 2020 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Entrypoint script for backend agnostic compile."""
+
+import os
+import pathlib
+import shutil
+import subprocess
+import sys
+import typing as T
+
+from . import mlog
+
+if T.TYPE_CHECKING:
+    import argparse
+
+
+def add_arguments(parser: 'argparse.ArgumentParser') -> None:
+    """Add compile specific arguments."""
+    parser.add_argument(
+        '-j', '--jobs',
+        action='store',
+        default=0,
+        type=int,
+        help='The number of worker jobs to run (if supported). If the value is less than 1 the build program will guess.'
+    )
+    parser.add_argument(
+        '-l', '--load-average',
+        action='store',
+        default=0,
+        type=int,
+        help='The system load average to try to maintain (if supported)'
+    )
+    parser.add_argument(
+        '--debug',
+        action='store_true',
+        help='Print debug information for the meson compile command.'
+    )
+    parser.add_argument(
+        'builddir',
+        type=pathlib.Path,
+        nargs='?',
+        default='.',
+        help='The directory containing build files to be built.'
+    )
```

Comment:
I think it should be -C to be consistent with test and install commands

Reply:
Ah, I added that for meson test, but I always thought people disliked that syntax. In retrospect I dislike it too. So I can either add support to meson install and meson test to not use -C as well (ie, both -C or a positional would work, but not at the same time), or I can add it here, preferences?

Reply:
`-C` should remain for the compatibility sake, if nothing else (there are tools that use a positional instead as well, but -C usage is pretty widespread). I have no opinion regarding positional arg though.

Reply:
I'm not a big fan of -C, but current status is a mess:
- setup: positional
- configure: positional
- dist: -C
- install: -C 
- introspect: positional
- init: --builddir
- test: -C

The trend seems to be that old commands use positional but new commands use -C. I think -C is pretty common (e.g. git, ninja, make) and it has the advantage of being more explicit in the case your command has other positional arguments. `meson compile foo bar`, does it build the `bar` target in `foo` builddir or the opposite?

Reply:
I don't have a strong opinion, but I'll commit to make the all of the commands do the same thing (without remove the current behavior of course), so, opinions on what the "right" thing do do is? @xclaesse, @nirbheek, @jpakkane